File: ts/src/Repository.ts
export interface Model {}

export type WithId<TModel extends Model, TId> = TModel & {
  _id: TId;
};

export type StringId<TModel extends Model> = TModel & {
  _id: string;
};

export type IdType<T> = T extends WithId<any, infer TId> ? TId : unknown;

export interface ListResult<T> {
  count: number;
  data: T[];
}

export interface Page {
  limit?: number;
  offset?: number;
}

export type Sort<T> = {
  [Property in keyof T]?: 1 | 0 | -1;
};

export interface Filter {}

export interface Repository<TModel extends Model, TFilter extends Filter> {
  getList(
    filter: TFilter,
    page?: Page,
    sort?: Sort<TModel>
  ): Promise<ListResult<StringId<TModel>>>;

  getById(id: string): Promise<StringId<TModel> | null>;

  create(data: TModel): Promise<string>;

  update(id: string, data: StringId<TModel>): Promise<boolean>;

  delete(id: string): Promise<void>;
}

File: go/repository.go
package repository

import "context"

type Repository[T Valid, Q Query] interface {
	GetById[T]
	GetList[T, Q]
	Create[T]
	Replace[T]
	Update[T]
	Delete
}

type GetById[T Valid] interface {
	GetById(ctx context.Context, id string) (T, error)
}

type Query map[string]interface{}

type Page struct {
	Limit  int
	Offset int
}

type Sort struct {
	Key   string
	Order bool
}

type GetList[T Valid, Q Query] interface {
	GetList(ctx context.Context, query Query, page Page, sort ...Sort) (ListResult[T], error)
}

type Create[T Valid] interface {
	Create(ctx context.Context, data T) (string, error)
}

type Replace[T Valid] interface {
	Replace(ctx context.Context, id string, data T) error
}

type Update[T Valid] interface {
	Update(ctx context.Context, id string, data T) (bool, error)
}

type Delete interface {
	Delete(ctx context.Context, id string) (bool, error)
}
